### PR TITLE
feat: remove extra allocations, add pred/lbl types to cb_explore

### PR DESCRIPTION
**Pull request by bassmang**
Addresses #3704

### CODE REVIEW

File: vowpalwabbit/reductions/cb/cb_explore.cc [hunk 1]
```diff
@@ -58,6 +58,7 @@ struct cb_explore
   VW::io::logger logger;
 
   size_t counter = 0;
+  int indexing = 1;
```
Comment (jackgerrits) [inline vowpalwabbit/reductions/cb/cb_explore.cc:61]: nit: `uint32_t`

File: vowpalwabbit/reductions/cb/cb_explore.cc [hunk 2]
```diff
@@ -83,12 +84,10 @@ void predict_or_learn_first(cb_explore& data, single_learner& base, example& ec)
   }
   else
   {
-    uint32_t chosen = ec.pred.multiclass - 1;
+    uint32_t chosen = (data.indexing == 1) ? ec.pred.multiclass - 1 : ec.pred.multiclass;
```
Comment (jackgerrits) [inline vowpalwabbit/reductions/cb/cb_explore.cc:87]: Could this just become
```suggestion
    uint32_t chosen = ec.pred.multiclass - data.indexing;
```

File: vowpalwabbit/reductions/cb/cb_explore.cc [hunk 3]
```diff
@@ -97,31 +96,30 @@ void predict_or_learn_greedy(cb_explore& data, single_learner& base, example& ec
...
   // pre-allocate pdf
 
   VW_DBG(ec) << "cb_explore: " << (is_learn ? "learn() " : "predict() ") << multiclass_pred_to_string(ec) << endl;
 
   probs.reserve(data.cbcs.num_actions);
   for (uint32_t i = 0; i < data.cbcs.num_actions; i++) probs.push_back({i, 0});
-  generate_epsilon_greedy(data.epsilon, ec.pred.multiclass - 1, begin_scores(probs), end_scores(probs));
-
-  ec.pred.a_s = probs;
+  uint32_t chosen = (data.indexing == 1) ? ec.pred.multiclass - 1 : ec.pred.multiclass;
```
Comment (jackgerrits) [inline vowpalwabbit/reductions/cb/cb_explore.cc:114]:
```suggestion
  uint32_t chosen = ec.pred.multiclass - data.indexing;
```

File: vowpalwabbit/reductions/cb/cb_explore.cc [hunk 4]
```diff
@@ -427,8 +426,10 @@ base_learner* cb_explore_setup(VW::setup_base_i& stack_builder)
   auto* l = make_reduction_learner(
       std::move(data), base, learn_ptr, predict_ptr, stack_builder.get_setupfn_name(cb_explore_setup) + name_addition)
                 .set_params_per_weight(params_per_weight)
+                .set_output_prediction_type(VW::prediction_type_t::multiclass)
```
Comment (jackgerrits) [inline vowpalwabbit/reductions/cb/cb_explore.cc:429]: This conflicts with the line just below

File: vowpalwabbit/reductions/cb/cb_explore.cc (hunk 1 above)
Comment (bassmang) [reply vowpalwabbit/reductions/cb/cb_explore.cc:61]: Removing indexing from this PR

File: vowpalwabbit/reductions/cb/cb_explore.cc (hunk 3 above)
Comment (bassmang) [reply vowpalwabbit/reductions/cb/cb_explore.cc:114]: Removing indexing from this PR

File: vowpalwabbit/reductions/cb/cb_explore.cc (hunk 2 above)
Comment (bassmang) [reply vowpalwabbit/reductions/cb/cb_explore.cc:87]: Removing indexing from this PR

File: vowpalwabbit/reductions/cb/cb_explore.cc (hunk 4 above)
Comment (bassmang) [reply vowpalwabbit/reductions/cb/cb_explore.cc:429]: Good catch! meant input_prediction_type here

**Review by jackgerrits — APPROVED** [review]

**Pull request merged**